// @antv/gui/esm/time-line/button.js
/**
 * @file 播放轴组件按钮
 * @author blackganglion
 */
import { __extends } from "tslib";
import { Group } from '@antv/g';
import * as _ from '@antv/util';
var Button = /** @class */ (function (_super) {
    __extends(Button, _super);
    function Button(cfg) {
        var _this = _super.call(this) || this;
        _this.config = _.deepMix({}, cfg);
        _this.init();
        return _this;
    }
    Button.prototype.update = function (cfg) {
        this.config = _.deepMix({}, this.config, cfg);
        this.updateElement();
        this.renderMarker();
    };
    Button.prototype.init = function () {
        this.initElement();
        this.renderMarker();
    };
    Button.prototype.initElement = function () {
        this.circle = this.addShape('circle', {
            attrs: {
                x: this.config.x,
                y: this.config.y,
                r: this.config.r,
                fill: '#607889',
            },
        });
        this.startMarker = this.addShape('path', {
            attrs: {
                path: this.getStartMarkerPath(),
                fill: '#ffffff',
            },
        });
        this.pauseGroupMarker = new Group();
        var width = (1 / 4) * this.config.r;
        var height = 0.5 * this.config.r * Math.sqrt(3);
        this.pauseLeftMarker = this.pauseGroupMarker.addShape('rect', {
            attrs: {
                x: this.config.x - (1 / 4 + 1 / 8) * this.config.r,
                y: this.config.y - height / 2,
                width: width,
                height: height,
                fill: '#ffffff',
            },
        });
        this.pauseRightMarker = this.pauseGroupMarker.addShape('rect', {
            attrs: {
                x: this.config.x + (1 / 8) * this.config.r,
                y: this.config.y - height / 2,
                width: width,
                height: height,
                fill: '#ffffff',
            },
        });
        this.add(this.pauseGroupMarker);
    };
    Button.prototype.updateElement = function () {
        this.circle.attr('x', this.config.x);
        this.circle.attr('y', this.config.y);
        this.circle.attr('r', this.config.r);
        this.startMarker.attr('path', this.getStartMarkerPath());
        var width = (1 / 4) * this.config.r;
        var height = 0.5 * this.config.r * Math.sqrt(3);
        this.pauseLeftMarker.attr('x', this.config.x - (1 / 4 + 1 / 8) * this.config.r);
        this.pauseLeftMarker.attr('y', this.config.y - height / 2);
        this.pauseLeftMarker.attr('width', width);
        this.pauseLeftMarker.attr('height', height);
        this.pauseRightMarker.attr('x', this.config.x + (1 / 8) * this.config.r);
        this.pauseRightMarker.attr('y', this.config.y - height / 2);
        this.pauseRightMarker.attr('width', width);
        this.pauseRightMarker.attr('height', height);
    };
    Button.prototype.renderMarker = function () {
        if (this.config.isPlay) {
            this.startMarker.hide();
            this.pauseGroupMarker.show();
        }
        else {
            this.startMarker.show();
            this.pauseGroupMarker.hide();
        }
    };
    /** 获取播放键 marker path */
    Button.prototype.getStartMarkerPath = function () {
        var sideLength = 0.5 * this.config.r * Math.sqrt(3);
        return [
            ['M', this.config.x - sideLength / Math.sqrt(3) / 2, this.config.y - sideLength / 2],
            ['L', this.config.x + sideLength / Math.sqrt(3), this.config.y],
            ['L', this.config.x - sideLength / Math.sqrt(3) / 2, this.config.y + sideLength / 2],
        ];
    };
    return Button;
}(Group));
export default Button;
